commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/core/EuclideanNormAlgorithms.java

```diff
@@ -0,0 +1,751 @@
+                return Math.sqrt(sum + c2);
+            }
+            return Math.sqrt(s3) * SCALE_DOWN;
+        }
+
+        /** Compute the square of {@code x}.
+         * @param x input value
+         * @return square of {@code x}
+         */
+        private static double square(final double x) {
+            return x * x;
+        }
+    }
+
+    /** Euclidean norm computation algorithm that uses {@link LinearCombinations} to perform
+     * an extended precision summation.
+     */
+    static final class ExtendedPrecisionLinearCombination implements ToDoubleFunction<double[]> {
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
+            // Find the magnitude limits ignoring zero
+            double max = 0;
+            double min = Double.POSITIVE_INFINITY;
+            for (int i = 0; i < v.length; i++) {
+                final double x = Math.abs(v[i]);
+                if (Double.isNaN(x)) {
+                    return x;
+                } else if (x > max) {
+                    max = x;
+                } else if (x < min && x != 0) {
+                    min = x;
+                }
+            }
+            // Edge cases
+            if (max == 0 || max == Double.POSITIVE_INFINITY) {
+                return max;
+            }
+            // Use scaling if required
+            double[] x = v;
+            double rescale = 1;
+            if (max > 0x1.0p500) {
+                // Too big so scale down
+                x = x.clone();
+                for (int i = 0; i < x.length; i++) {
+                    x[i] *= 0x1.0p-600;
+                }
+                rescale = 0x1.0p600;
+            } else if (min < 0x1.0p-500 && max < 0x1.0p100) {
+                // Too small so scale up
+                x = x.clone();
+                for (int i = 0; i < x.length; i++) {
+                    x[i] *= 0x1.0p600;
+                }
+                rescale = 0x1.0p-600;
+            }
+            return Math.sqrt(Sum.ofProducts(x, x).getAsDouble()) * rescale;
+        }
+    }
+
+    /** Modification of {@link ExtendedPrecisionLinearCombination} that uses an optimized version of the
+     * linear combination computation.
+     */
+    static final class ExtendedPrecisionLinearCombinationMod implements ToDoubleFunction<double[]> {
+
+        /** Threshold for scaling small numbers. */
+        private static final double SMALL_THRESH = 0x1.0p-500;
+
+        /** Threshold for scaling large numbers. */
+        private static final double LARGE_THRESH = 0x1.0p+500;
+
+        /** Value used to scale down large numbers. */
+        private static final double SCALE_DOWN = 0x1.0p-600;
+
+        /** Value used to scale up small numbers. */
+        private static final double SCALE_UP = 0x1.0p+600;
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 436-457 into a separate method.